### PR TITLE
os/bluestore: fix bluefs log run out of space(improved)

### PR DESCRIPTION
**Pull request by ifed01**
This is an evolution of #41888. The primary driver for this patch was a desire to eliminate the need to perform redundant encodings during the initial size estimation in the initial patch. 
Additionally new patch hopefully handles a corner case : still there was no100% guarantee that the runway allocation provides enough space for the expanded log file metadata - potentially that metadata could grow above the estimated runway due to the allocation.

Fixes: https://tracker.ceph.com/issues/51217
Signed-off-by: Igor Fedotov <ifedotov@suse.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-07-21_08:19:42-rados-wip-kefu-testing-2021-07-21-1257-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-21_13:52:36-rados-wip-kefu-testing-2021-07-21-2006-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-07-21_14:42:35-rados-wip-kefu-testing-2021-07-21-2006-distro-basic-smithi/

e2e failures tracked by https://tracker.ceph.com/issues/51728

File: src/test/objectstore/store_test.cc [hunk 1]
```diff
@@ -9139,6 +9139,22 @@ TEST_P(StoreTestSpecificAUSize, SpilloverFixed3Test) {
...
+    return;
+
+  const char* m[][10] = {
+    { "bluestore_min_alloc_size", "8192", 0 }, // must be the first!
+    { "bluefs_alloc_size", "8192", 0 },
+    { "bluefs_max_prefetch ", "8192", 0 },
+    { "bluefs_min_log_runway ", "8192", 0 },
+    { "bluefs_max_log_runway ", "8192", 0 },
+    { "bluefs_log_compact_min_size ", "21474836480", 0 },
+    { 0 },
```
Comment (yushu20171007) [inline src/test/objectstore/store_test.cc:9153]: maybe need set 'bluefs_shared_alloc_size=8192' after Ceph Nautilus

**Review by aclamk — CHANGES_REQUESTED** [review]
Reference on extending BlueFS log:
https://github.com/ceph/ceph/pull/41888#issuecomment-891634398

I think we should not add partial fixes but try to fix issue completely.
My idea for fix would be:
1) always keep X bytes free for BlueFS log
2) if transaction we are about to write reduces free space below X extend log *before* writing to it
3) log extension is done by injecting extra transaction with only ONE op, and update to INO 1. This bluefs_transaction_t is allowed to have the same id as next transaction (real transaction)
4) We track value of X as size of transaction we needed to update INO 1

Pros:
Should fix BlueFS extension problems
Cons:
We will have 2 transactions in log with same id. (current _replay already accepts it)

**Comment by ifed01** [comment]
> Reference on extending BlueFS log:
> [#41888 (comment)](https://github.com/ceph/ceph/pull/41888#issuecomment-891634398)
> 
> I think we should not add partial fixes but try to fix issue completely.
Yeah, agree. Realized that a couple days ago as well which inspired me to refactor BlueFS log implementation.
Here is what I've got as a result:
https://github.com/ceph/ceph/pull/42612

This still lacks some functionality for volume migration and probably needs a more elegant split between legacy and new functionality, e.g. base BlueFS class + two descendants with some member overrides.


> My idea for fix would be:
> 
> 1. always keep X bytes free for BlueFS log
> 2. if transaction we are about to write reduces free space below X extend log _before_ writing to it
> 3. log extension is done by injecting extra transaction with only ONE op, and update to INO 1. This bluefs_transaction_t is allowed to have the same id as next transaction (real transaction)
> 4. We track value of X as size of transaction we needed to update INO 1
IMO this still provides no 100% guarantee that the upcoming log extension fits into X, e.g. due to a fact that such an extension actually depends on the new transaction itself. Not to mention that I'm a bit scared of introducing significant changes into current BlueFS log implementation ;)

> 
> Pros:
> Should fix BlueFS extension problems
> Cons:
> We will have 2 transactions in log with same id. (current _replay already accepts it)

**Pull request closed without merging**